File: src/backend/langboard/core/ai/BotRunner.py
from typing import Any
from ..utils.decorators import class_instance, singleton
from .BaseBot import BaseBot
from .BotResponse import LangflowStreamResponse
from .InternalBotType import InternalBotType


@class_instance()
@singleton
class BotRunner:
    def __init__(self):
        self.__bot_factory: dict[InternalBotType, BaseBot] = {}

    async def run(self, bot_type: InternalBotType, data: dict[str, Any]) -> str | LangflowStreamResponse | None:
        bot = self.__get_bot(bot_type)
        if bot is None:
            return None
        return await bot.run(data)

    async def run_abortable(
        self, bot_type: InternalBotType, data: dict[str, Any], task_id: str
    ) -> str | LangflowStreamResponse | None:
        bot = self.__get_bot(bot_type)
        if bot is None:
            return None
        return await bot.run_abortable(data, task_id)

    async def abort(self, bot_type: InternalBotType, task_id: str):
        bot = self.__get_bot(bot_type)
        if bot is None:
            return
        await bot.abort(task_id)

    async def upload_file(self, bot_type: InternalBotType, file: Any) -> str | None:
        bot = self.__get_bot(bot_type)
        if bot is None:
            return None
        return await bot.upload_file(file)

    async def is_available(self, bot_type: InternalBotType) -> bool:
        if bot_type not in BaseBot.__bots__:
            return False

        if bot_type not in self.__bot_factory:
            self.__bot_factory[bot_type] = BaseBot.__bots__[bot_type]()

        return await self.__bot_factory[bot_type].is_available()

    def __get_bot(self, bot_type: InternalBotType) -> BaseBot | None:
        if bot_type not in BaseBot.__bots__:
            return None

        if bot_type not in self.__bot_factory:
            self.__bot_factory[bot_type] = BaseBot.__bots__[bot_type]()

        return self.__bot_factory[bot_type]

File: src/socket/src/core/ai/BotRunner.ts
import { getBot, IBotIsAvailableOptions, IBotRunAbortableOptions, IBotRunOptions, IBotUploadOptions } from "@/core/ai/BaseBot";
import ISocketClient from "@/core/server/ISocketClient";
import { EInternalBotType } from "@/models/InternalBot";
import { ESocketTopic, GLOBAL_TOPIC_ID } from "@langboard/core/enums";

export interface IBotAbortOptions {
    botType: EInternalBotType;
    taskID: string;
    client?: ISocketClient;
}

class BotRunner {
    public static async run({ internalBot, data }: IBotRunOptions) {
        const bot = getBot(internalBot.bot_type);
        if (!bot) {
            return null;
        }
        return await bot.run({ internalBot, data });
    }

    public static async runAbortable({ internalBot, ...options }: IBotRunAbortableOptions) {
        const bot = getBot(internalBot.bot_type);
        if (!bot) {
            return null;
        }
        return await bot.runAbortable({ internalBot, ...options });
    }

    public static async createTitle({ internalBot, data }: IBotRunOptions) {
        const bot = getBot(internalBot.bot_type);
        if (!bot) {
            return null;
        }
        return await bot.createTitle({ internalBot, data });
    }

    public static async abort({ botType, taskID, client }: IBotAbortOptions): Promise<void> {
        const bot = getBot(botType);
        if (!bot) {
            return;
        }
        await bot.abort(taskID);
        client?.send({
            topic: ESocketTopic.Global,
            topic_id: GLOBAL_TOPIC_ID,
            event: "task:aborted",
            data: {
                task_id: taskID,
            },
        });
    }

    public static async isAvailable({ internalBot }: IBotIsAvailableOptions): Promise<bool> {
        const bot = getBot(internalBot.bot_type);
        if (!bot) {
            return false;
        }
        return await bot.isAvailable({ internalBot });
    }

    public static async upload({ internalBot, file }: IBotUploadOptions): Promise<string | null> {
        const bot = getBot(internalBot.bot_type);
        if (!bot) {
            return null;
        }
        return await bot.upload({ internalBot, file });
    }

    public static isAborted(botType: EInternalBotType, task_id: string): bool {
        const bot = getBot(botType);
        if (!bot) {
            return true;
        }
        return bot.isAborted(task_id);
    }

    public static createAbortedChecker(botType: EInternalBotType, task_id: string) {
        const isAborted = () => {
            const aborted = BotRunner.isAborted(botType, task_id);
            return aborted;
        };
        return isAborted;
    }
}

export default BotRunner;
